### ISSUE TITLE
[improvement] react/no-unstable-nested-components & false negatives

### ISSUE THREAD

**Issue by xgbuils** [jsx-eslint/eslint-plugin-react#3622]
A component is allowed to be defined inside other component if it is created by a hook or a HOC. For example:

``` JavaScript
function ParentComponent() {
  const InsideDefinedFunctionComponent = withHappiness(OutsideDefinedFunctionComponent);
  return (
    <div>
      <InsideDefinedFunctionComponent />
    </div>
  );
}
```

I don't have much knowdledge about the tree that generates eslint. However, I'm thinking in a rule that maybe can help to solve these false negatives:

1. Get an instance of a component (it's just a Identifier called by a jsx function). 
2. Traverse the parent nodes of this node and find a function/method declaration
  a. If there is no function/method the validation succeeds (it gets the global scope without finding an internal declaration of the component). 
  b. Otherwise continue.
3. Check the variable declaration or parameters of the function
  a. If the variables or parameters have the same name as the identifier, then, the validation fails. 
  b. Otherwise go to 2.

It's true that this rule is not failing when the internal declared component is not called:

``` JavaScript
function ParentComponent() {
  const InsideDefinedFunctionComponent = () => <div/>;
  return (
    <div>
      Unused InsideDefinedFunctionComponent
    </div>
  );
}
```

 However I think it's not an issue, because a internal declared component which is not used cannot produce any reconciliation issue. Then is a false positive. Additionally the rule I'm proposing is catching errors like that:

``` JavaScript
function tryingToHackRule() {
  const InsideDefinedFunctionComponent = () => <div/>; // Error! declaration of an instantiated component inside of a function
  return (
    <div>
      <InsideDefinedFunctionComponent />
    </div>
  );
}

function ParentComponent() {
  return tryingToHackRule();
}
```

**Comment by ljharb** [comment]
(from #3621)

Of course it can produce a reconciliation issue.

First of all, it's a very strange thing to use HOCs with hooks, since hooks were created partially to get rid of HOCs.

Second, if you *must* create it inside the component, that's what `useMemo` is for - memoize it, so that you can avoid the real problem that otherwise exists. If you do this, I'd expect the rule to be OK with it.

**Comment by xgbuils** [comment]
> Of course it can produce a reconciliation issue.

Could you explain me how this code can produce a reconciliation issue?

```JavaScript
function ParentComponent() {
  const InsideDefinedFunctionComponent = () => <div/>;
  return (
    <div>
      Unused InsideDefinedFunctionComponent
    </div>
  );
}
```

> First of all, it's a very strange thing to use HOCs with hooks, since hooks were created partially to get rid of HOCs.

In my example I didn't considered an example of HOCs with hooks and I guess this rule intend to be supported by HOCs, hooks, functional or class components, right?

> Second, if you must create it inside the component, that's what useMemo is for - memoize it, so that you can avoid the real problem that otherwise exists. If you do this, I'd expect the rule to be OK with it.

Of course issues can be solved without using eslint. But I assume that eslint rules intend to catch issues automatically without human intervention. On the other hand, not always who sets the rule is the same who should follow the rule. For this reason, I think if there is a possibility to improve this rule, maybe could be considered.

Additionally in the [documentation of the rule](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-unstable-nested-components.md) is recommending to not use `useCallback` or `useMemo` because can be refactored in another way:

> ⚠️ WARNING ⚠️:
>
> Creating nested but memoized components is currently not detected by this rule but should also be avoided. If the useCallback or useMemo hook has no dependency, you can safely move the component definition out of the render function. If the hook does have dependencies, you should refactor the code so that you're able to move the component definition out of the render function.

Thanks for the review.

**Comment by ljharb** [comment]
I agree that in your last example, it's unused, so it should just be deleted. The reconciliation issue is if you render it inside an element that might care about component identity.

**Comment by ljharb** [comment]
I think this is answered; this does not seem like a false negative to me.

**Comment by xgbuils** [comment]
> I agree that in your last example, it's unused, so it should just be deleted.

Yep. It's something that I said in my first post (and repeated in my last post).

>  The reconciliation issue is if you render it inside an element that might care about component identity.

In fact, I think this is the key point of the rule I proposed. Meanwhile the current rule is checking  if a component is declared inside another component, my proposed rule gets de jsx element and checks if the name of jsx element is declared in the function above or not.

> I think this is answered; this does not seem like a false negative to me.

I'm not sure what you think is not a false negative: A component inside useMemo or useCallback? In this case I don't understand why in the documentation and tests are commented as false negatives.
If you mean the declared  but not rendered component(what I think it is a false _positive_), then, OK. But my proposal it's not just about removing this false positive. It's about proposing another algorithm able to detect false negatives like useMemo, useCallback and HOC that creates components inside other components.

I'm feeling my proposal has not been read well because I had to clarify things that I didn't say in my first post or repeat things I did say in my first post. Additionally, I've seen part of considerations just was omitted.

I guess thar mantaining and open source project(or multiples) is hard and sometimes it's difficult to read and write carefully. Then, I'm gonna try to implement my no-unstable-nested-component proposal rule  by myself in another project using this project test cases as a base. And if I think it's valuable I'll reopen this issue.

**Comment by ljharb** [comment]
Defining any component at "not the top level of the module" (which works fine for HOCs) is a problem, and that's what the rule reports on.

**Comment by xgbuils** [comment]
I'm implementing my own eslint rule and I think can be more easier than I proposed at first step:

1. Get a JSXElement node and save the node as `currentNode` and the name as `componentName`.
2. Get the parent of `currentNode`.
2.1. If the node is a FunctionDeclaration/MethodDefinition/ArrowExpr<ession go to **3.**
2.2. If the node is a BlockStatement get the identifiers of variable, function or class declarations.
2.2.1. If `componentName` is included in the identifiers list, show an error.
2.2.2. Otherwise, go to **2**.
3. Get the identifiers of params of the function or the method
3.1.  If `componentName` is included in the identifiers list, show an error.
3.2. Otherwise is valid.

In this way, all the variants(with class, with function...) of this code snippet:
```JavaScript
const Parent = () => {
  const Nested = () => <div/>;
  return <Nested />;
};
```
are well reported.

HOC creating component inside component is reported (not reported in this library):
```JavaScript
const ParentComponent = () => {
const NestedComponent = withHappiness(SadNestedComponent);
  return <NestedComponent />;
};
```

HOC creating component outside the component is not reported:
```JavaScript
const NestedComponent = withHappiness(SadNestedComponent);
const ParentComponent = () => {
  return <NestedComponent />;
};
```

Using `useMemo` or `useCallback` is reported (not reported in this library):
```JavaScript
const ParentComponent = () => {
  const NestedComponent = useMemo(() => <Component />, []);
  return <NestedComponent />;
};
```

Additionally is more robust catching this kind of issues (this library is not able to detect them):
```JavaScript
const renderMethod = () => {
  const NestedComponent = () => <Component />;
  return <NestedComponent />;
};
class ParentComponent {
  render() {
    renderMethod();
  }
}
```

My proposed algorithm does not catch components as props. This is a bit controversial, but I think this library is not taking the good approach. This library is allowing renderProps if starts with prefix `render`:

```JavaScript
const ParentComponent = () => {
  return <Component renderProp={() => <NestedComponent/>} />;
};
```
but no one knows how this renderProp is used in Component definition. `renderProp` could be managed in this way:

```JavaScript
const Component = ({renderProp: ComponentProp}) => {
  return <ComponentProp />;
};
```
Then, checking if a component props have components and reporting depending on a naming convention is a weak check. I think it's better checking the Component implementation because the real root of the reconcilation problem is defining a component in a way that can receive components as prop. My proposal is catching this issue in the point **3** of the algorithm. In fact:
```JavaScript
const Component = ({renderProp: ComponentProp}) => {
  return <ComponentProp />;
};
```
is causing a reconcilition issue. However:

```JavaScript
const Component = ({renderProp: ComponentProp}) => {
  return <div>{renderProp()}</div>;
};
```
is not causing any issue. My proposal is managing this kind of issues in the correct way.

Cheers.
